// @abc/protonpack/src/components/image/balancedImage/example.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Row, Col, BalancedImage} from '../../../export'

export default class BalancedImageExample extends Component {
    render() {
        const colStyleLight = {
            alignItems: 'flex-start',
            background:'#fff',
            color: '#4a4a4a',
            padding: '1em 0',
            textAlign: 'center',
        }
        const colStyleDark = {
            alignItems: 'flex-start',
            background:'#4a4a4a',
            color: '#fff',
            padding: '1em 0',
            textAlign: 'center',
        }
        return (
            <div className="exampleSection">
                <h3 className="mb-3">Balanced Image</h3>
                <Row style={colStyleLight}>
                    <Col>
                        <BalancedImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" round />
                    </Col>
                    <Col>
                        <BalancedImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" />
                    </Col>
                    <Col>
                        <BalancedImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" round hasBorder/>
                        <div>w/Border</div>
                    </Col>
                    <Col>
                        <BalancedImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" hasBorder/>
                        <div>w/Border</div>
                    </Col>
                </Row>
                <Row style={colStyleDark}>
                    <Col>
                        <BalancedImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" round />
                    </Col>
                    <Col>
                        <BalancedImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" />
                    </Col>
                    <Col>
                        <BalancedImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" round hasBorder/>
                        <div>w/Border</div>
                    </Col>
                    <Col>
                        <BalancedImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" hasBorder/>
                        <div>w/Border</div>
                    </Col>
                </Row>
            </div>
        )
    }
}

BalancedImageExample.propTypes = {
    className: PropTypes.string,
}
