// @abc/fedagents/src/components/stepBar/example.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Row, Col, Button} from '@abc/protonpack'
import {StepBar, IconGroupLink} from '../../export'

const titles = [
    'The First Step',
    (<IconGroupLink type="check" onClick={() => alert('I\'m Clickable')}>Accepts Components Too</IconGroupLink>), // eslint-disable-line no-alert
    'The Almost Fourth Step',
    'The Next to Last Step',
    'The Final Step',
]

export default class StepBarExample extends Component {
    constructor() {
        super()

        this.handleNextStep = this.handleNextStep.bind(this)
        this.handlePrevStep = this.handlePrevStep.bind(this)

        this.state = {
            step: 1,
        }
    }

    handleNextStep() {
        this.setState({
            step: Math.min(titles.length, this.state.step + 1),
        })
    }

    handlePrevStep() {
        this.setState({
            step: Math.max(1,this.state.step - 1),
        })
    }

    render() {
        return (
            <div className={this.props.className}>
                <h2>Step Bar</h2>
                <Row className="mb-4">
                    <Col>
                        <StepBar step={this.state.step} titles={titles}/>
                    </Col>
                </Row>
                <h4>Step Bar w/o Step Numbers</h4>
                <Row className="mb-4">
                    <Col>
                        <StepBar step={this.state.step} titles={titles} hideStepNumber/>
                    </Col>
                </Row>
                <Row className="mb-4">
                    <Col>
                        <Button size="sm" btnStyle={this.state.step <= 1 ? 'default' : 'primary'} disabled={this.state.step <= 1} onClick={this.handlePrevStep} style={{float:'left'}}>Prev Step</Button>
                        <Button size="sm" btnStyle={this.state.step >= titles.length ? 'default' : 'primary'} disabled={this.state.step >= titles.length} onClick={this.handleNextStep} style={{float:'right'}}>Next Step</Button>
                    </Col>
                </Row>
            </div>
        )
    }
}

StepBarExample.propTypes = {
    className: PropTypes.string,
}
