// @abc/fedagents/src/components/accordion/example.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import Accordion from './'
import {Row, Col, ToggleButton} from '@abc/protonpack'

export default class AccordionExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            exampleOpen: true,
            doop: [
                {
                    name: 'Amazonia',
                    checked: false,
                },
                {
                    name: 'Amphibios 9',
                    checked: false,
                },
                {
                    name: 'Chapek 9',
                    checked: false,
                },
                {
                    name: 'Decapod 10',
                    checked: false,
                },
                {
                    name:  'Earth',
                    checked: false,
                },
                {
                    name: 'Neptune',
                    checked: false,
                },
                {
                    name:  'Neutral Planet',
                    checked: false,
                },
                {
                    name:  'Nylar IV',
                    checked: false,
                },
                {
                    name: 'Trisol',
                    checked: false,
                },
                {
                    name: 'Wormulon',
                    checked: false,
                },
            ],
        }

        this.handleToggle = this.handleToggle.bind(this)
    }

    demoClick(e) {
        e.stopPropagation()
        alert('clicked more text!') // eslint-disable-line no-alert
    }

    toggleButton(i) {
        const newDoopList = this.state.doop
        newDoopList[i].checked = !this.state.doop[i].checked

        this.setState({
            doop: [...newDoopList],
        })
    }

    handleToggle(open) {
        this.setState({exampleOpen: open})
    }

    render() {
        const selectListContent = (
            <div style={{display: 'flex', width: '100%', lineHeight: '2.8125em', justifyContent: 'space-between'}}>
                <div>
                    Selectable List
                </div>
                <div onClick={this.demoClick}>
                    more text
                </div>
            </div>
        )

        return (
            <div className={this.props.className}>
                <h2>Accordion</h2>
                <Row>
                    <Col>
                        <Accordion displayContent="Expand Me!">
                            This is an Accordion
                        </Accordion>
                    </Col>
                </Row>
                <Row>
                    <Col>
                        <Accordion
                            displayContent="Preopened"
                            isOpen={this.state.exampleOpen}
                            onToggle={this.handleToggle}
                        >
                            already open!
                        </Accordion>
                    </Col>
                </Row>
                <Row>
                    <Col>
                        <Accordion displayContent={selectListContent}>
                            {this.state.doop.map((doop, i) => (
                                <div key={i}>
                                    <ToggleButton
                                        btnStyle="primary"
                                        block
                                        checked={doop.checked}
                                        onChange={() => this.toggleButton(i)}
                                        key={`${doop.name}-${i}`}
                                        style={{width: '100%'}}
                                    >
                                        <span>{doop.name}</span>
                                    </ToggleButton>
                                </div>
                            ))}
                        </Accordion>
                    </Col>
                </Row>
            </div>
        )
    }
}

AccordionExample.propTypes = {
    className: PropTypes.string,
}
