// @abc/stay-puft-js/index.js
const Eureka = require('@abc/eureka-js-client')

const PUFT_SERVICE_ID = 'stay-puft'
const blobUrlBase = 'userPhoto'
const locationURLBase = 'locationDocument'
const locationLogoURLBase = 'locationLogo'
const saveCardDocument = 'saveCardOnFileTerms'

class StayPuftService extends Eureka {

    constructor(logger) {
        super({logger})
        this.start()

        this.getBlob = this.getBlob.bind(this)
    }

    getBlob(blobId, accessToken) {
        const options = {url: `/${blobUrlBase}/${blobId}`, method: 'get', auth: {bearer: accessToken}}

        return this.callEurekaService(PUFT_SERVICE_ID, options, `Problem getting Blob ${blobId}`)
    }

    /**
     * Save blob as binary array
     *
     * @see dev-stay-puft.greatestclubmanagementsoftwareever.com/#resources-puft-create-user-photo
     */
    saveBlob(metadata, content, accessToken) {
        const options = {
            url: `/${blobUrlBase}`,
            method: 'post',
            json: {
                metadata,
                content,
            },
            auth: {bearer: accessToken},
        }

        return this.callEurekaService(PUFT_SERVICE_ID, options, `Problem saving Blob ${metadata.id}`)
    }

    getInstruction(textId, accessToken) {
        const options = {url: `/${locationURLBase}/${textId}`, method: 'get', auth: {bearer: accessToken}}

        return this.callEurekaService(PUFT_SERVICE_ID, options, `Problem getting text ${textId}`)
    }

    /**
     * Get location logo by Id
     *
     * @see dev-stay-puft.greatestclubmanagementsoftwareever.com/#resources-puft-get-location-logo
     */
    getLocationLogo({accessToken, logoId}) {
        const options = {url: `/${locationLogoURLBase}/${logoId}`, method: 'get', auth: {bearer: accessToken}}

        return this.callEurekaService(PUFT_SERVICE_ID, options, `Problem getting logo ${logoId}`)
    }

    /**
     * Get the 'Save Card on File Terms document'
     *
     * @see http://dev-stay-puft.greatestclubmanagementsoftwareever.com/#resources-puft-create-static-document
     */
    getSaveCardTermsDocument() {
        const options = {url: `/static/${saveCardDocument}`, method: 'get'}

        return this.callEurekaService(PUFT_SERVICE_ID, options, 'Problem getting "Save Card on File Terms document"')
    }
}

module.exports = StayPuftService
