// @abc/protonpack/src/components/image/eventImage/example.js
import React from 'react'
import PropTypes from 'prop-types'
import {Row, Col, EventImage, SecondaryFont} from '../../../export'

const EventImageExample = () => {
    const colStyleLight = {
        alignItems: 'center',
        background:'#fff',
        color: '#4a4a4a',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'flex-start',
        padding: '1em 0',
        textAlign: 'center',
    }
    const colStyleDark = {
        alignItems: 'center',
        background:'#4a4a4a',
        color: '#fff',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'flex-start',
        padding: '1em 0',
        textAlign: 'center',
    }
    return (
        <div className="exampleSection">
            <h3 className="mb-3">Event Image</h3>
            <Row>
                <Col style={colStyleLight}>
                    <EventImage src="./bogus.png" size="4.375rem"/>
                    <div>Default Color</div>
                </Col>
                <Col style={colStyleLight}>
                    <EventImage src="./bogus.png" size="4.375rem" hasBorder/>
                    <div>Default Color - w/Border</div>
                </Col>
                <Col style={colStyleLight}>
                    <div>
                        <SecondaryFont color="primary">
                            <EventImage src="./bogus.png" size="4.375rem" inheritColor/>
                        </SecondaryFont>
                        <div>Inherit Color</div>
                    </div>
                </Col>
                <Col style={colStyleLight}>
                    <div>
                        <SecondaryFont color="primary">
                            <EventImage src="./bogus.png" size="4.375rem" inheritColor hasBorder/>
                        </SecondaryFont>
                        <div>Inherit Color - w/Border</div>
                    </div>
                </Col>
            </Row>
            <Row>
                <Col style={colStyleDark}>
                    <EventImage src="./bogus.png" size="4.375rem"/>
                    <div>Default Color</div>
                </Col>
                <Col style={colStyleDark}>
                    <EventImage src="./bogus.png" size="4.375rem" hasBorder/>
                    <div>Default Color - w/Border</div>
                </Col>
                <Col style={colStyleDark}>
                    <div>
                        <SecondaryFont color="primary">
                            <EventImage src="./bogus.png" size="4.375rem" inheritColor/>
                        </SecondaryFont>
                        <div>Inherit Color</div>
                    </div>
                </Col>
                <Col style={colStyleDark}>
                    <div>
                        <SecondaryFont color="primary">
                            <EventImage src="./bogus.png" size="4.375rem" inheritColor hasBorder/>
                        </SecondaryFont>
                        <div>Inherit Color - w/Border</div>
                    </div>
                </Col>
            </Row>
        </div>
    )
}

EventImageExample.propTypes = {
    className: PropTypes.string,
}

export default EventImageExample
