// @abc/protonpack/src/components/divider/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'

import {Row, Col} from '../layout/export'
import classnames from 'classnames'
import makeClassnames from '../../utils/bem'

export class Divider extends Component {
    buildClasses() {
        const {type, direction, color, fullWidth, className} = this.props
        const classes = makeClassnames({
            element: `divider-${type}`,
            modifiers: {
                element: {
                    direction,
                    color,
                    fullWidth,
                },
            },
        })

        if (className && className !== '') {
            classes[className] = true
        }

        return classes
    }

    render() {
        const {
            type,
            direction, // eslint-disable-line no-unused-vars
            color, // eslint-disable-line no-unused-vars
            fullWidth, // eslint-disable-line no-unused-vars
            className, // eslint-disable-line no-unused-vars
            children,
            ['data-abc-id']: dataAbcId,
            ...remain
        } = this.props

        const abcId = dataAbcId || `${type}Divider`
        const classes = this.buildClasses()

        let divider
        if (type === 'text') {
            divider = (
                <div data-abc-id={abcId} className={classnames(classes)} {...remain}>
                    {children}
                </div>
            )
        } else {
            if (direction === 'vertical') {
                divider = (
                    <span className={classnames(classes)} data-abc-id={abcId} />
                )
            } else {
                divider = (
                    <Row xs={{justifyContent: 'center'}}>
                        <Col {...remain}>
                            <hr className={classnames(classes)} data-abc-id={abcId} />
                        </Col>
                    </Row>
                )
            }
        }
        return divider
    }
}

Divider.defaultProps = {
    type: 'border',
    direction: 'horizontal',
    color: 'gray-lighter',
}

Divider.propTypes = {
    type: PropTypes.oneOf([
        'text',
        'border',
    ]),
    color: PropTypes.oneOf([
        'primary',
        'secondary',
        'tertiary',
        'gray-lighter',
        'gray-light',
        'gray-dark',
        'white',
        'success',
        'danger',
        'warning',
    ]),
    direction: PropTypes.oneOf([
        'horizontal',
        'vertical',
    ]),
    className: PropTypes.string,
    children: PropTypes.node,
    fullWidth: PropTypes.bool,
}

export const FullWidthDivider = () => <Divider fullWidth />
export const VerticalDivider = () => <Divider direction='vertical' />
