// @abc/protonpack/src/brander/common/range/BranderRange.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import Range from './'

export default class BranderRange extends Component {
    constructor(props) {
        super(props)

        this.state = {
            value: parseInt(String(this.props.styles.getPropertyValue(this.props.CSSProperty)).trim(), 10),
        }

        this.handleChange = this.handleChange.bind(this)
    }

    handleChange(e) {
        const {value} = e.target

        this.setState({value})
        document.documentElement.style.setProperty(this.props.CSSProperty, `${value}px`)
    }

    render() {
        // eslint-disable-next-line no-unused-vars
        const {CSSProperty, styles, ...rest} = this.props

        return (
            <Range
                onChange={this.handleChange}
                value={this.state.value}
                {...rest}
            />
        )
    }
}

BranderRange.propTypes = {
    CSSProperty: PropTypes.string.isRequired,
    styles: PropTypes.object.isRequired,
}
