// @abc/protonpack/src/components/typography/baseText/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import makeClassnames from '../../../utils/bem'
import classnames from 'classnames'

export default class BaseText extends Component {
    get elementName() {
        return 'base-font'
    }

    buildClasses() {
        const {color, size, weight, children} = this.props
        const className = children.props.className
        const classes = makeClassnames({
            block: 'font-style',
            element: this.elementName,
            modifiers: {
                block: {
                    color,
                    weight,
                },
                element: {
                    size,
                },
            },
        })

        if (className && className !== '') {
            classes[className] = true
        }

        return classes
    }

    render() {
        const {children, alpha} = this.props

        const classes = this.buildClasses()
        const style = children.props.style || {}
        style.opacity = alpha

        return React.cloneElement(children, {className: classnames(classes), style})
    }
}

BaseText.propTypes = {
    children: PropTypes.node.isRequired,
    color: PropTypes.oneOf([
        'primary',
        'secondary',
        'tertiary',
        'gray-lightest',
        'gray-lighter',
        'gray-light',
        'gray-dark',
        'white',
        'success',
        'danger',
        'warning',
    ]),
    size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl', 'xxl']),
    weight: PropTypes.oneOf(['light', 'normal', 'medium', 'bold']),
    alpha: PropTypes.number,
}

BaseText.defaultProps = {
    size: 'lg',
    color: 'gray-dark',
    weight: 'bold',
}
