// @abc/protonpack/src/examples/AlertExample.js
import React, {Component} from 'react'
import {Alert, Button} from '../export'

export default class AlertExample extends Component {
    constructor(props) {
        super(props)

        this.defaultState = {
            successVisible: true,
            warningVisible: true,
            infoVisible: true,
            dangerVisible: true,
        }
        this.state = this.defaultState
    }

    hideAlert(type) {
        const nextState = {}
        nextState[`${type}Visible`] = false
        this.setState(nextState)
    }

    resetAlertState() {
        this.setState(this.defaultState)
    }

    render() {
        return (
            <div className="exampleSection">
                <h2 className="mb-2">Alerts</h2>

                <Alert type="success" isOpen><strong>Success</strong> Clamps</Alert>
                <Alert type="info" isOpen><strong>Info</strong> Donbot</Alert>
                <Alert type="warning" isOpen><strong>Warning</strong> Flexo</Alert>
                <Alert type="danger" isOpen><strong>Danger</strong> Hedonismbot</Alert>

                <div className="exampleSubsection">
                    <h3 className="mb-1">
                        Dissmissable Alerts  <Button size="sm" btnStyle="secondary" onClick={() => {
                            this.resetAlertState()
                        }}>Reset</Button>
                    </h3>
                </div>

                <Alert type="success" isOpen={this.state.successVisible} onDismiss={() => this.hideAlert('success')}>
                    I'm a <strong>SUCCESS</strong> Message. Click the &times; to dismiss me.
                </Alert>
                <Alert type="info" isOpen={this.state.infoVisible} onDismiss={() => this.hideAlert('info')}>
                    I'm an <strong>INFO</strong> Message. Click the &times; to dismiss me.
                </Alert>
                <Alert type="warning" isOpen={this.state.warningVisible} onDismiss={() => this.hideAlert('warning')}>
                    I'm a <strong>WARNING</strong> Message. Click the &times; to dismiss me.
                </Alert>
                <Alert type="danger" isOpen={this.state.dangerVisible} onDismiss={() => this.hideAlert('danger')}>
                    I'm a <strong>DANGER</strong> Message. Click the &times; to dismiss me.
                </Alert>
            </div>
        )
    }
}
