// @abc/protonpack/src/examples/FormExample.js
import React, {Component} from 'react'
import DatePickerExample from '../components/formElements/datepicker/example'
import ToggleButtonExample from '../components/formElements/toggleButton/example'
import ReactSelectExample from '../components/formElements/reactSelect/example'
import {
    Form,
    FormSection,
    FormGroup,
    Label,
    LabelGroup,
    Input,
    Textarea,
    ToggleSwitch,
    ReactSelect,
    Row,
    Col,
    Button,
    Checkbox,
    Radio,
} from '../export'

export default class FormExample extends Component {
    constructor(props) {
        super(props)
        this.state = {
            reactSelectValue: '',
        }
    }

    render() {
        return (
            <div className="exampleSection">
                <h2 className="mb-2">Forms</h2>
                <Form>
                    <FormSection>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Input" id="regularInput">
                                    <Input type="text" data-abc-id="caitlinTest"/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="noBorder Input" id="noBorderInput">
                                    <Input type="text" data-abc-id="caitlinTest" noBorder/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Disabled Input" id="disabledInput">
                                    <Input type="text" defaultValue="readOnly Input" readOnly/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Hidden Label" id="hiddenLabelInput" hideLabel>
                                    <Input type="text" placeholder="Hidden Label"/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                    </FormSection>
                    <FormSection>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Disabled" hideLabel>
                                    <Input
                                        type="text"
                                        placeholder="Search"
                                        icon={{type: 'search', position: 'left'}}
                                        disabled
                                    />
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Input" id="iconInput">
                                    <Input
                                        type="text"
                                        placeholder="Hidden Label"
                                        icon={{type: 'eye', position: 'right'}}
                                    />
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Icon noBorder" id="iconInputnoBorder">
                                    <Input
                                        type="text"
                                        placeholder="Hidden Label"
                                        icon={{type: 'eye', position: 'right'}}
                                        noBorder
                                    />
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Standalone Label" id="standaloneLabelInput" standaloneLabel>
                                    <Input type="text"/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                    </FormSection>

                    <FormSection>
                        <Col>
                            <ReactSelectExample/>
                        </Col>
                    </FormSection>

                    <FormSection>
                        <Col>
                            <h4>Check Box</h4>
                            <div className="form-check">
                                <div className="pt-2 pb-2 d-flex">
                                    <LabelGroup
                                        check
                                        standaloneLabel
                                        className="test"
                                        inline="reverse"
                                        label="Remember Me With Placeholder"
                                        id="rememberMePlaceholder"
                                    >
                                        <Checkbox/>
                                    </LabelGroup>
                                </div>
                                <div className="pt-2 pb-2 d-flex">
                                    <LabelGroup
                                        check
                                        standaloneLabel
                                        inline
                                        label="Remember Me Disabled"
                                        id="rememberMePlaceholderDisabled"
                                    >
                                        <Checkbox disabled checked/>
                                    </LabelGroup>
                                </div>
                            </div>
                        </Col>
                        <Col>
                            <h4>Radio</h4>
                            <div className="pt-2 pb-2">
                                <Radio name="radiogroup" id="bender" label="Bender" checked/>
                            </div>

                            <div className="pt-2 pb-2">
                                <Radio name="radiogroup" id="flexo" label="Flexo"/>
                            </div>

                            <div className="pt-2 pb-2">
                                <Radio name="radiogroup" id="disabled-radio" label="Disabled Radio" disabled/>
                            </div>

                            <div className="pt-2 pb-2">
                                <Radio
                                    name="disabled-group"
                                    id="disabled-checked-radio"
                                    label="Disabled Checked Radio"
                                    disabled
                                    checked
                                />
                            </div>
                        </Col>
                        <Col>
                            <h4>Toggle Switch</h4>
                            <div className="form-check">
                                <div className="pt-2 pb-2 no-line-height">
                                    <ToggleSwitch/>
                                </div>

                                <div className="pt-2 pb-2 no-line-height">
                                    <ToggleSwitch disabled/>
                                </div>

                                <div className="pt-2 pb-2 no-line-height">
                                    <ToggleSwitch checked disabled/>
                                </div>
                            </div>
                        </Col>
                    </FormSection>

                    <FormSection>
                        <Col>
                            <ToggleButtonExample/>
                        </Col>
                    </FormSection>

                    <FormSection>
                        <Col>
                            <FormGroup
                                statusMessage={{
                                    status: 'success',
                                    text: "Our crew is replacable, your package isn't.",
                                }}
                            >
                                <LabelGroup label="Success State" id="successStateInput">
                                    <Input type="text"/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup statusMessage={{status: 'warning', text: 'Not Bender'}}>
                                <LabelGroup label="Warning State" id="warningStateInput">
                                    <Input
                                        type="text"
                                        placeholder="enter robot"
                                        defaultValue="Flexo"
                                    />
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup statusMessage={{status: 'danger', text: 'Do Not Eat!'}}>
                                <LabelGroup label="Danger State" id="dangerStateInput">
                                    <Input
                                        type="text"
                                        placeholder="pizza"
                                        defaultValue="Popplers"
                                    />
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                    </FormSection>

                    <FormSection>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Textarea" id="regularTextarea">
                                    <Textarea data-abc-id="textareaId"/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Disabled Textarea" id="disableTextarea">
                                    <Textarea defaultValue="readOnly Textarea" readOnly/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup>
                                <LabelGroup label="Hidden Textarea Label" id="hiddenLabelTextarea" hideLabel>
                                    <Textarea placeholder="Textarea with hidden Label"/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                    </FormSection>

                    <FormSection>
                        <Col>
                            <FormGroup statusMessage={{status: 'success', text: 'Textarea with success msg.'}}>
                                <LabelGroup label="Success Textarea State" id="successStateTextarea">
                                    <Textarea/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup statusMessage={{status: 'warning', text: 'Textarea with warning msg.'}}>
                                <LabelGroup label="Warning Textarea State" id="warningStateTextarea">
                                    <Textarea/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup statusMessage={{status: 'danger', text: 'Textarea with error msg.'}}>
                                <LabelGroup label="Danger Textarea State" id="dangerStateTextarea">
                                    <Textarea/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                    </FormSection>

                    <FormSection>
                        <Col>
                            <LabelGroup label="Without FormGroup as Parent" id="withoutFormGroup">
                                <Textarea/>
                            </LabelGroup>
                        </Col>
                    </FormSection>

                    <FormSection>
                        <Col>
                            <FormGroup>
                                <LabelGroup
                                    label="Standalone Textarea Label"
                                    id="standaloneLabelTextarea"
                                    standaloneLabel
                                >
                                    <Textarea/>
                                </LabelGroup>
                            </FormGroup>
                        </Col>
                    </FormSection>

                    <h4>Date Picker</h4>

                    <FormSection>
                        <Col>
                            <DatePickerExample/>
                        </Col>
                    </FormSection>

                    <Row xs={{justifyContent: 'center'}}>
                        <Col xs={6} md={4} lg={3} xl={2}>
                            <Button block>CANCEL</Button>
                        </Col>
                        <Col xs={6} md={4} lg={3} xl={2}>
                            <Button block btnStyle="primary">SAVE</Button>
                        </Col>
                    </Row>
                </Form>

                <Row className="mb-1">
                    <Col>
                        <h3>Inline Form</h3>
                        <Form inline onSubmit={false}>
                            <FormGroup>
                                <LabelGroup label="User:" id="userInput" standaloneLabel>
                                    <Input type="text" placeholder="deliveryboy42"/>
                                </LabelGroup>
                            </FormGroup>
                            <FormGroup>
                                <LabelGroup
                                    label="Password:"
                                    id="passwordInput"
                                    standaloneLabel
                                >
                                    <Input type="password"/>
                                </LabelGroup>
                            </FormGroup>
                            <Button btnStyle="primary">Submit</Button>
                        </Form>
                    </Col>
                </Row>

                <Row className="mb-1">
                    <Col>
                        <h3>Micro Form</h3>
                        <Form onSubmit={false}>
                            <Row sm={{justifyContent: 'center'}}>
                                <Col sm={3}>
                                    <FormSection size="sm">
                                        <Col sm={12}>
                                            <FormGroup size="sm">
                                                <LabelGroup label="Address 1" id="address1" hideLabel>
                                                    <Input type="text" placeholder="Address 1" size="sm"/>
                                                </LabelGroup>
                                            </FormGroup>
                                        </Col>
                                        <Col sm={12}>
                                            <FormGroup size="sm">
                                                <LabelGroup label="Address 2" id="address2" hideLabel>
                                                    <Input type="text" placeholder="Address 2" size="sm"/>
                                                </LabelGroup>
                                            </FormGroup>
                                        </Col>
                                        <Col sm={12}>
                                            <FormGroup size="sm">
                                                <LabelGroup label="City" id="city" hideLabel>
                                                    <Input type="text" placeholder="City" size="sm"/>
                                                </LabelGroup>
                                            </FormGroup>
                                        </Col>
                                        <Col sm={{col: 5, noGutters: 'right'}}>
                                            <FormGroup size="sm">
                                                <Label srOnly>State</Label>
                                                <ReactSelect
                                                    size="sm"
                                                    placeholder="State"
                                                    options={[
                                                        {value: 'AR', label: 'Arkansas'},
                                                        {value: 'AK', label: 'Alaska'},
                                                        {value: 'AL', label: 'Alabama'},
                                                        {value: 'etc', label: 'etc.'},
                                                    ]}
                                                    clearable={false}
                                                    value={this.state.reactSelectValue}
                                                    onChange={this.handleChange}
                                                />
                                            </FormGroup>
                                        </Col>
                                        <Col sm={7}>
                                            <FormGroup size="sm">
                                                <LabelGroup label="Zip Code" id="zipCode" hideLabel>
                                                    <Input type="text" placeholder="Zip Code" size="sm"/>
                                                </LabelGroup>
                                            </FormGroup>
                                        </Col>
                                    </FormSection>
                                    <Row>
                                        <Col>
                                            <Button block type="submit" btnStyle="primary" size="sm">SUBMIT</Button>
                                        </Col>
                                    </Row>
                                </Col>
                            </Row>
                        </Form>
                    </Col>
                </Row>
            </div>
        )
    }
}
