// @abc/protonpack/src/examples/SwiperExample.js
import React, {Component} from 'react'
import {Col, Row, Swiper, Card} from '../export'

const items = [
    'First Card',
    'Second Card',
    'Third Card',
    'Fourth Card',
    'Fifth Card',
    'Sixth Card',
]

const clickableItems = [
    {id: 14, name: 'First Click Card'},
    {id: 23, name: 'Second Click Card'},
    {id: 31, name: 'Third Click Card'},
    {id: 42, name: 'Fourth Click Card'},
    {id: 53, name: 'Fifth Click Card'},
    {id: 63, name: 'Sixth Click Card'},
]

export default class SwiperExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            clickItemIndex: 0,
        }

        this.onItemClick = this.onItemClick.bind(this)
        this.handleSlideChange = this.handleSlideChange.bind(this)
    }

    onItemClick(index) {
        this.setState({clickItemIndex: index})
    }

    handleSlideChange(swiper) {
        const {clickItemIndex} = this.state

        if (swiper.activeIndex !== clickItemIndex) {
            this.setState({clickItemIndex: swiper.activeIndex})
        }
    }

    render() {
        return (
            <div className="exampleSection">
                <h2 className="mb-2">Swipers</h2>

                <Row className="mb-3 exampleSubsection">
                    <Col>
                        <h3 className="mb-3">Default</h3>
                        <Swiper>
                            {items.map((item, index) => (
                                <Card key={index} backgroundColor="lightGray">
                                    {item}
                                </Card>
                            ))}
                        </Swiper>
                    </Col>
                </Row>

                <Row className="mb-3 exampleSubsection">
                    <Col>
                        <h3 className="mb-3">With Ratio - 16:9</h3>
                        <Swiper ratio="16:9">
                            {items.map((item, index) => (
                                <Card key={index} backgroundColor="lightGray">
                                    {item}
                                </Card>
                            ))}
                        </Swiper>
                    </Col>
                </Row>

                <Row className="mb-3 exampleSubsection">
                    <Col>
                        <h3 className="mb-3">With Centered, Spaced, and Clickable Items</h3>
                        <Swiper
                            slidesPerView="auto"
                            centeredSlides={true}
                            spaceBetween={16}
                            activeIndex={this.state.clickItemIndex}
                            onTransitionStart={this.handleSlideChange}
                        >
                            {clickableItems.map((item, index) => (
                                <Card
                                    key={item.id}
                                    backgroundColor="lightGray"
                                    onClick={this.onItemClick.bind(this, index)}
                                >
                                    {item.name}
                                </Card>
                            ))}
                        </Swiper>
                    </Col>
                </Row>

            </div>
        )
    }
}
