// @abc/fedagents/src/components/reduxFormGroup/index.js
import React from 'react'
import PropTypes from 'prop-types'
import {FormGroup} from '@abc/protonpack'

const ReduxFormGroup = (props) => {
    // eslint-disable-next-line no-unused-vars
    const {input, children, meta, label, ...remain} = props
    const showError = meta.error && meta.touched

    const statusMessage = {}
    if (showError) {
        statusMessage.status = 'danger'
        statusMessage.text = meta.error
    }

    return (
        <FormGroup statusMessage={statusMessage} {...remain} >
            {
                React.Children.map(children, child => React.cloneElement(child, {
                    ...input,
                    id: child.props.id || input.name,
                    size: remain.size || child.props.size,
                    disabled: meta.submitting || child.props.disabled,
                }))
            }
        </FormGroup>
    )
}

ReduxFormGroup.propTypes = {
    input: PropTypes.object.isRequired,
    meta: PropTypes.object.isRequired,
    children: PropTypes.oneOfType([
        PropTypes.object,
        PropTypes.array,
    ]).isRequired,
    label: PropTypes.string,
}

export default ReduxFormGroup
