// @abc/protonpack/src/components/modal/orchestrator/example/slide.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Button, Modal, ModalOrchestrator} from '../../../../export'

export default class ModalOrchestratorExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            currentIndex: 0,
            isOpen: false,
        }

        this.toggleOpen = this.toggleOpen.bind(this)
    }

    goBack() {
        this.setState({
            currentIndex: this.state.currentIndex - 1,
        })
    }

    goNext() {
        this.setState({
            currentIndex: this.state.currentIndex + 1,
        })
    }

    toggleOpen() {
        this.setState({
            isOpen: !this.state.isOpen,
            currentIndex: 0,
        })
    }

    render() {
        const firstButtons = [
            {
                text: 'Cancel',
                props: {
                    btnStyle: 'primary',
                    onClick: this.toggleOpen,
                },
            },
            {
                text: 'Next',
                props: {
                    onClick: () => this.goNext(),
                },
            },
        ]

        const middleButtons = [
            {
                text: 'Back',
                props: {
                    btnStyle: 'primary',
                    onClick: () => this.goBack(),
                },
            },
            {
                text: 'Next',
                props: {
                    onClick: () => this.goNext(),
                },
            },
        ]

        const lastButtons = [
            {
                text: 'Back',
                props: {
                    btnStyle: 'primary',
                    onClick: () => this.goBack(),
                },
            },
            {
                text: 'Finish',
                props: {
                    onClick: this.toggleOpen,
                },
            },
        ]

        return (
            <div>
                <Button btnStyle="primary" block onClick={() => this.toggleOpen()}>Slide Modal Orchestrator</Button>
                <ModalOrchestrator
                    currentIndex={this.state.currentIndex}
                    isOpen={this.state.isOpen}
                    toggle={this.toggleOpen}
                    transitionType="slide"
                >
                    <Modal banner={{title: 'The First Modal'}} buttons={firstButtons} toggle={() => 'This toggle will be overwritten'}>
                        Press next to advance to the next modal.
                    </Modal>
                    <Modal banner={{title: 'The Second Modal'}} buttons={middleButtons} toggle={() => 'This toggle will be overwritten'}>
                        Press next to advance, or back to go to the previous modal.
                    </Modal>
                    <Modal banner={{title: 'The Last Modal'}} buttons={lastButtons}  toggle={() => 'This toggle will be overwritten'}>
                        Press back to go to the previous modal.
                    </Modal>
                </ModalOrchestrator>
            </div>
        )
    }
}

ModalOrchestratorExample.propTypes = {
    className: PropTypes.string,
}
