// @abc/fedagents/src/components/email/display/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import styles from './display.styl'

export default class Email extends Component {
    getEmail() {
        const {displayData} = this.props
        const email = (typeof displayData === 'object') ? (displayData && displayData.address) : displayData

        return email && email.toLowerCase()
    }

    emailClass(className) {
        return className ? `${styles.email} ${className}` : styles.email
    }

    getEmailLink(remain, email) {
        const {children, showAddress, className} = this.props
        let emailLink

        if (children) {
            emailLink = children
        } else if (showAddress) {
            emailLink = email
        } else {
            emailLink = 'Email'
        }

        return <a href={`mailto:${email}`} className={this.emailClass(className)} {...remain}>{emailLink}</a>
    }

    render() {
        // eslint-disable-next-line
        let {displayData, showAddress, noLink, children, className, ...remain} = this.props

        // eslint-disable-next-line
        remain = {...remain, 'data-abc-id': 'email'}

        const email = this.getEmail()
        let emailDisplay

        if (email) {
            if (noLink) {
                emailDisplay = <span className={this.emailClass(className)} {...remain}>{email}</span>
            } else {
                emailDisplay = this.getEmailLink(remain, email)
            }
        } else {
            emailDisplay = <em className={this.emailClass(className)} {...remain}>none provided</em>
        }

        return emailDisplay
    }
}

Email.propTypes = {
    displayData: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.object,
    ]),
    showAddress: PropTypes.bool,
    noLink: PropTypes.bool,
    children: PropTypes.node,
    className: PropTypes.string,
}
