// @abc/protonpack/src/components/alert/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import Icon from '@abc/quarkicons'

export default class Alert extends Component {
    buildClasses(type) {
        const classes = 'alert'
        const alertType = type || 'danger'

        return `${classes} alert-${alertType}`
    }

    render() {
        // eslint-disable-next-line no-unused-vars
        const {onDismiss, children, type, isOpen, className, ...remain} = this.props

        const alert = (
            <div
                className={this.buildClasses(type)}
                role="alert"
                data-abc-id="alert"
                {...remain}
            >
                {onDismiss &&
                    <button className="close" aria-label="Close" onClick={onDismiss}>
                        <Icon type='x-sm'/>
                    </button>
                }

                {children}
            </div>
        )

        return isOpen && alert
    }
}

Alert.propTypes = {
    className: PropTypes.string,
    children: PropTypes.node.isRequired,
    type: PropTypes.string,
    isOpen: PropTypes.bool.isRequired,
    onDismiss: PropTypes.func,
}
