// @abc/fedagents/src/components/loading/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import Icon from '@abc/quarkicons'
import styles from './loading.styl'

export default class Loading extends Component {
    constructor(props) {
        super(props)
        this.state = {hidden: styles.hidden}
        this.show = this.show.bind(this)
        this.buildClasses = this.buildClasses.bind(this)
    }

    componentDidMount() {
        this.timeout = setTimeout(() => {
            this.show()
        }, this.props.timeout)
    }

    componentWillUnmount() {
        clearTimeout(this.timeout)
        this.timeout = undefined
    }

    buildClasses() {
        const classes = []
        this.props.className && classes.push(this.props.className)
        this.state.hidden ? classes.push(this.state.hidden) : classes.push(styles.center)

        if (this.props.sm) {
            classes.push(styles.loaderSm)
        } else if (this.props.lg) {
            classes.push(styles.loaderLg)
        } else {
            classes.push(styles.loaderMd)
        }

        return classes.join(' ').trim()
    }

    show() {
        this.setState({hidden: null})
    }

    render() {
        return (
            <div className={this.buildClasses()} data-abc-id="loadingSpinner">
                <Icon type="spinner" className={styles.spinner} />
                <span className="sr-only" data-abc-id="loadingText">Loading...</span>
            </div>
        )
    }
}

Loading.propTypes = {
    timeout: PropTypes.number,
    center: PropTypes.bool,
    className: PropTypes.string,
    sm: PropTypes.bool,
    md: PropTypes.bool,
    lg: PropTypes.bool, 
}

Loading.defaultProps = {
    timeout: 500,
}
