// @abc/fedagents/webpack.prod.config.js
const path = require('path')
const HtmlWebpackPlugin = require('html-webpack-plugin')

module.exports = {
    bail: true,
    entry: ['./src/export.js'],
    output: {
        path: path.join(__dirname, 'dist'),
        filename: 'fedagents.js',
        library: 'fedagents',
        libraryTarget: 'umd',
    },
    externals: [
        {
            react: {
                root: 'React',
                commonjs2: 'react',
                commonjs: 'react',
                amd: 'react',
            },
            'react-dom': {
                root: 'ReactDOM',
                commonjs2: 'react-dom',
                commonjs: 'react-dom',
                amd: 'react-dom',
            },
            'react-router': {
                root: 'ReactRouter',
                commonjs2: 'react-router',
                commonjs: 'react-router',
                amd: 'react-router',
            },
            '@abc/protonpack': {
                root: 'ProtonPack',
                commonjs2: '@abc/protonpack',
                commonjs: '@abc/protonpack',
                amd: '@abc/protonpack',
            },
            '@abc/quarkicons': {
                root: '@abc/quarkicons',
                commonjs2: '@abc/quarkicons',
                commonjs: '@abc/quarkicons',
                amd: '@abc/quarkicons',
            },
        },
    ],
    devServer: {},
    devTool: '',
    plugins: [
        new HtmlWebpackPlugin({template: './src/index.html'}),
    ],
}
