// @abc/protonpack/src/components/banner/modal/example.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Row,Col} from '../../../export'

export default class BannerModalExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            modal: false,
            modalSubtitle: false,
            modalIndicator: false,
        }

        this.toggle = this.toggle.bind(this)
        this.toggleIndicator = this.toggleIndicator.bind(this)
        this.toggleSubtitle = this.toggleSubtitle.bind(this)
    }

    toggle() {
        this.setState({
            modal: !this.state.modal,
        })
    }

    toggleIndicator() {
        this.setState({
            modalIndicator: !this.state.modalIndicator,
        })
    }

    toggleSubtitle() {
        this.setState({
            modalSubtitle: !this.state.modalSubtitle,
        })
    }

    render() {
        return (
            <div className="mt-3 mb-2">
                <h3>Modal Banners</h3>
                <Row className="mt-3 mb-2">
                    <Col xs={12} sm={6}>
                        See Modal Examples
                    </Col>
                </Row>
            </div>
        )
    }
}

BannerModalExample.propTypes = {
    className: PropTypes.string,
}
