// @abc/protonpack/src/components/banner/modal/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import {SecondaryFont} from '../../typography/export'
import Indicator, {iconMap} from '../../indicator'

export default class BannerModal extends Component {
    render() {
        const {indicator, title, children, className, styleType, ...remain} = this.props

        const classes = classnames('banner--modal', className)

        return (
            <div className={classes} data-abc-id="bannerModal" {...remain}>
                {indicator &&
                    <Indicator type={indicator} className="banner--modal__indicator" styleType={styleType} />
                }
                <h1>{title}</h1>
                <SecondaryFont color="gray-light">
                    <div data-abc-id="bannerModalContent" className="banner--modal__content">
                        {children}
                    </div>
                </SecondaryFont>
            </div>
        )
    }
}

BannerModal.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    indicator: PropTypes.oneOf(Object.keys(iconMap)),
    styleType: PropTypes.string,
    title: PropTypes.string.isRequired,
}
