// @abc/protonpack/src/components/layout/container/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'

export default class Container extends Component {
    buildClasses(fluid, fluidHeight, className) {
        const classes = []

        if (fluid) {
            classes.push('container-fluid')
        } else {
            classes.push('container')
        }

        if (className) {
            classes.push(className)
        }

        if (fluidHeight) {
            classes.push('fluid-height')
        }

        return classes.join(' ')
    }

    render() {
        const {fluid, className, fluidHeight, ...remain} = this.props

        const classNames = this.buildClasses(fluid, fluidHeight, className)

        return (
            <div {...remain} className={classNames}>
                {this.props.children}
            </div>
        )
    }
}

Container.propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    fluid: PropTypes.bool,
    top: PropTypes.string,
    middle: PropTypes.string,
    bottom: PropTypes.string,
    fluidHeight: PropTypes.bool,
}
