// @abc/protonpack/src/components/cards/card/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import makeClassnames from '../../../utils/bem'
import classnames from 'classnames'


export default class Card extends Component {
    get elementName() {
        return 'card'
    }

    buildClasses() {
        const {className, backgroundColor, removePadding, noShadow, noBorder} = this.props
        const classes = makeClassnames({
            element: this.elementName,
            modifiers: {
                element: {
                    backgroundColor,
                    removePadding,
                    noShadow,
                    noBorder,
                },
            },
        })

        if (className) {
            classes[className] = true
        }

        return classes
    }

    render() {
        // eslint-disable-next-line no-unused-vars
        const {children, backgroundColor, className, removePadding, noBorder, noShadow, ...remain} = this.props
        const classes = classnames(this.buildClasses(this.props))

        return <div className={classes} data-abc-id="card" {...remain}>{children}</div>
    }
}

Card.propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    backgroundColor: PropTypes.oneOf(['lightGray', 'lighterGray', 'white']),
    removePadding: PropTypes.bool,
    noShadow: PropTypes.bool,
    noBorder: PropTypes.bool,
}
