// @abc/protonpack/src/components/formElements/textarea/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'

export default class Textarea extends Component {
    constructor(props) {
        super(props)

        this.state = {
            focused: false,
        }

        this.handleFocus = this.handleFocus.bind(this)
        this.handleBlur = this.handleBlur.bind(this)
    }

    componentDidUpdate(prevProps) {
        if (!this.props.value && prevProps.value && !this.state.focused) {
            this.props.onClear && this.props.onClear()
        }
    }

    handleFocus() {
        this.setState({focused: true})
        this.props.onFocus()
    }

    handleBlur(e) {
        this.setState({focused: false})
        this.props.onBlur(e)
    }

    buildClasses() {
        const {statusMessage, size, className} = this.props
        const classes = []

        classes.push('form-control')

        if (size) {
            classes.push(`form-control-${size}`)
        }

        if (statusMessage && statusMessage.status) {
            classes.push(`form-control-${statusMessage.status}`)
        }

        if (className) {
            classes.push(className)
        }

        return classes.join(' ')
    }

    render() {
        const {
            statusMessage, // eslint-disable-line no-unused-vars
            size, // eslint-disable-line no-unused-vars
            className, // eslint-disable-line no-unused-vars
            onClear, // eslint-disable-line no-unused-vars
            onFocus, // eslint-disable-line no-unused-vars
            onBlur, // eslint-disable-line no-unused-vars
            ['data-abc-id']: dataAbcId,
            ...formGroup
        } = this.props

        // eslint-disable-next-line no-unused-vars
        const {normalize, validate, onValidationFieldInitialized, onValidationFieldFailure, onValidationFieldSuccess, ...remain} = formGroup

        const classes = this.buildClasses()
        const abcId = dataAbcId || 'textarea'

        return (
            <textarea
                className={classes}
                data-abc-id={abcId}
                onFocus={this.handleFocus}
                onBlur={this.handleBlur}
                {...remain}
            />
        )
    }
}

Textarea.propTypes = {
    className: PropTypes.string,
    value: PropTypes.string,
    statusMessage: PropTypes.object,
    size: PropTypes.string,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    onClear: PropTypes.func,
}
