// @abc/fedagents/src/components/virtualized/list/example/scrollableContainer.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {VirtualizedList} from '../../../../export'

import range from 'lodash/range'
import uniqueId from 'lodash/uniqueId'

const makeNewListItem = (count) => {
    return {
        myId: uniqueId('virtualized_list_uid_'),
        number: count + 1,
    }
}

export default class ScrollableContainerVirtualizedListExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            count: 200,
            options: range(200).map(makeNewListItem),
        }

        this.renderGenericItem = this.renderGenericItem.bind(this)
    }

    renderGenericItem(test) {
        return <div className="example">Inside Container {test.number}</div>
    }

    render() {
        return (
            <div
                style={{overflow: 'auto', height: '350px'}}
            >
                <VirtualizedList
                    direction="horizontal"
                    displayData={this.state.options}
                    idKey="myId"
                    item={this.renderGenericItem}
                    {...this.props}
                />
            </div>
        )
    }
}

ScrollableContainerVirtualizedListExample.propTypes = {
    className: PropTypes.string,
}
