// @abc/protonpack/src/components/progressBar/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import makeClassnames from '../../utils/bem'
import classnames from 'classnames'

export default class ProgressBar extends Component {
    get elementName() {
        return 'progress-bar'
    }

    buildClasses() {
        const {block, barStyle, background, border, size, innerRadius} = this.props

        return {
            container: {
                ...makeClassnames({
                    block,
                    element: this.elementName,
                    modifiers: {
                        element: {
                            background,
                            border,
                            size,
                            style: barStyle,
                        },
                    },
                }),
            },
            bar: {
                ...makeClassnames({
                    block: this.elementName,
                    element: 'progress',
                    modifiers: {
                        element: {
                            innerRadius,
                        },
                    },
                }),
            },
        }
    }

    render() {
        // eslint-disable-next-line no-unused-vars
        const {className, block, barStyle, background, border, size, progress, ['data-abc-id']: dataAbcId, innerRadius, ...remain} = this.props

        const classes = this.buildClasses()

        const abcId = dataAbcId || 'progress-bar'
        const elementStyle = {
            width: `${progress * 100}%`,
        }

        return (
            <div
                className={classnames(classes.container,className)}
                data-abc-id={abcId}
                {...remain}
            >
                <div className={classnames(classes.bar)} style={elementStyle}/>
            </div>
        )
    }
}

ProgressBar.propTypes = {
    background: PropTypes.oneOf([
        'danger',
        'disabled',
        'info',
        'none',
        'primary',
        'secondary',
        'success',
        'warning',
    ]),
    barStyle: PropTypes.oneOf([
        'danger',
        'danger-light',
        'disabled',
        'info',
        'primary',
        'secondary',
        'success',
        'success-light',
        'warning',
    ]),
    block: PropTypes.string,
    border: PropTypes.bool,
    className: PropTypes.string,
    progress: PropTypes.number,
    size:  PropTypes.oneOf([
        'xs',
        'sm',
        'md',
        'lg',
        'xl',
    ]),
    innerRadius: PropTypes.bool,
}

ProgressBar.defaultProps = {
    progress: 1,
    size: 'md',
}
