// @abc/protonpack/src/components/editor/example.js
import React, {Component} from 'react'
import RichEditor from './'
import {Col, Row} from '../../export'
import './example.scss'

export default class RichEditorExample extends Component {
    constructor(props) {
        super(props)

        const data = '{"blocks":[{"key":"a7skh","text":"Hello","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"5uen7","text":"Bold Bold Bold","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":0,"length":14,"style":"BOLD"}],"entityRanges":[],"data":{}},{"key":"60o9o","text":"Italic Italic Itaclic","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":0,"length":21,"style":"ITALIC"}],"entityRanges":[],"data":{}}],"entityMap":{}}'

        const textdata = 'Hello customer!!!'
        this.state = {
            basicContent: '',
            textContent: textdata,
            content: data,
        }

        this.config = {
            placeholderText: 'Edit Content Here!',
            charCounterCount: false,
        }

        this.handleEditorChange = this.handleEditorChange.bind(this)
    }

    handleEditorChange(editorContent) {
        this.setState({basicContent: editorContent})
    }

    render() {
        return (
            <div>
                <h2 className="mb-2">Editor</h2>
                <Row className="mb-3" style={{justifyContent: 'center'}}>
                    <Col xs={10}>
                        <div className="mb-3" style={{textAlign: 'center', fontWeight: 'bold'}}> Basic Rich Editor </div>
                        <RichEditor
                            className='editorClass'
                            data-abc-id='rich-editor-1'
                            content={this.state.basicContent}
                            onChange={this.handleEditorChange}
                        />
                    </Col>
                </Row>
                <Row className="mb-3" style={{justifyContent: 'center'}}>
                    <Col xs={10}>
                        <div className="mb-3" style={{textAlign: 'center', fontWeight: 'bold'}}> Rich Editor - placeholder and w/default height</div>
                        <RichEditor
                            className='editorClass'
                            data-abc-id='rich-editor-2'
                            content={this.state.basicContent}
                            placeholder='My placeholder text...'
                            onChange={this.handleEditorChange}
                            height={200}
                        />
                    </Col>
                </Row>
                <Row className="mb-3" style={{justifyContent: 'center'}}>
                    <Col xs={10}>
                        <div className="mb-3" style={{textAlign: 'center', fontWeight: 'bold'}}> Rich Editor - ReadOnly W/text content </div>
                        <RichEditor
                            className="editorPadding"
                            data-abc-id='rich-editor-3'
                            content={this.state.textContent}
                            onChange={this.handleEditorChange}
                            readOnly
                        />
                    </Col>
                </Row>
                <Row className="mb-3" style={{justifyContent: 'center'}}>
                    <Col xs={10}>
                        <div className="mb-3" style={{textAlign: 'center', fontWeight: 'bold'}}> Rich Editor - ReadOnly W/formatted content </div>
                        <RichEditor
                            className="editorPadding"
                            data-abc-id='rich-editor-4'
                            content={this.state.content}
                            onChange={this.handleEditorChange}
                            size="md"
                            weight="bold"
                            readOnly
                        />
                    </Col>
                </Row>
            </div>
        )
    }
}
