// @abc/protonpack/src/components/formElements/radio/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import Label from '../label'
import Input from '../input'

export default class Radio extends Component {
    buildClasses() {
        const {className} = this.props
        const classes = ['form-check']

        if (className) {
            classes.push(className)
        }
        return classes.join(' ')
    }

    render() {
        // eslint-disable-next-line no-unused-vars
        const {id, name, label, hideLabel, className, ...formGroup} = this.props

        // eslint-disable-next-line no-unused-vars
        const {normalize, validate, onValidationFieldInitialized, onValidationFieldFailure, onValidationFieldSuccess, ...remain} = formGroup

        const classes = this.buildClasses()
        return (
            <div className={classes}>
                <Input id={id} name={name} type="radio" {...remain} />
                <Label htmlFor={id} srOnly={hideLabel} radio>{label}</Label>
            </div>
        )
    }
}

Radio.propTypes = {
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    hideLabel: PropTypes.bool,
    className: PropTypes.string,
}
