// @abc/protonpack/src/brander/form/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Row, Col} from 'layout/export'
import ColorPicker from '../colorPicker'
import {FormGroup, LabelGroup, ReactSelect, Label} from 'formElements/export'
import {formElements, labelStates} from '../data'
import BranderRange from '../common/range/BranderRange'
import './formBrander.scss'

class FormBrander extends Component {
    constructor(props) {
        super(props)

        this.state = {
            element: '',
            labelState: '',
        }
    }

    render() {
        const {styles} = this.props
        const {element, labelState} = this.state

        return (
            <Row xs={{justifyContent: 'center'}}>
                <Col sm={12}>
                    <ColorPicker
                        customProperty="--formBackground"
                        label="Background"
                        className="form-brander-background"
                        color={String(styles.getPropertyValue('--formBackground')).trim()}
                        isCSSProperty
                    />
                </Col>
                <Col sm={6}>
                    <ColorPicker
                        customProperty="--formText"
                        label="Text"
                        className="form-brander-text"
                        color={String(styles.getPropertyValue('--formText')).trim()}
                    />
                </Col>
                <Col sm={6}>
                    <FormGroup size="sm" style={{marginBottom: 0}}>
                        <LabelGroup label="Border Radius" standaloneLabel >
                            <BranderRange
                                id="formBorderRadius"
                                CSSProperty="--formBorderRadius"
                                styles={styles}
                                min={0}
                                max={30}
                            />
                        </LabelGroup>
                    </FormGroup>
                </Col>
                <Col sm={6}>
                    <FormGroup size="sm">
                        <Label>Element</Label>
                        <ReactSelect
                            size="sm"
                            options={formElements}
                            onChange={element => this.setState({element})}
                            value={element}
                            placeholder="Select One..."
                        />
                    </FormGroup>
                </Col>
                <Col sm={6}>
                    <FormGroup size="sm">
                        <Label>State</Label>
                        <ReactSelect
                            size="sm"
                            options={labelStates}
                            onChange={labelState => this.setState({labelState})}
                            value={labelState}
                        />
                    </FormGroup>
                </Col>
                {this.state.element &&
                    <Col sm={12}>
                        <Row>
                            <Col sm={12}>
                                <ColorPicker
                                    customProperty={`--form${element}${labelState}`}
                                    label="Color"
                                    className={`form-brander-${element.toLowerCase()}${labelState ? `-${labelState.toLowerCase()}-` : '-'}color`}
                                    color={String(styles.getPropertyValue(`--form${element}${labelState}`)).trim()}
                                />
                            </Col>
                        </Row>
                    </Col>
                }
            </Row>
        )
    }
}

FormBrander.propTypes = {
    styles: PropTypes.object.isRequired,
}

export default FormBrander
