// @abc/fedagents/src/components/banners/modalBanner/example.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Row, Col, Button, Modal, ModalHeader, ModalBody, ModalFooter, SecondaryFont, VerticalDivider} from '@abc/protonpack'
import {ModalBanner} from '../../../export'

export default class ModalBannerExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            modal: false,
            modalSubtitle: false,
            modalIndicator: false,
        }

        this.toggle = this.toggle.bind(this)
        this.toggleIndicator = this.toggleIndicator.bind(this)
        this.toggleSubtitle = this.toggleSubtitle.bind(this)
    }

    toggle() {
        this.setState({
            modal: !this.state.modal,
        })
    }

    toggleIndicator() {
        this.setState({
            modalIndicator: !this.state.modalIndicator,
        })
    }

    toggleSubtitle() {
        this.setState({
            modalSubtitle: !this.state.modalSubtitle,
        })
    }

    render() {
        return (
            <div className={this.props.className}>
                <h3>Modal Banners</h3>
                <Row className={this.props.className}>
                    <Col xs={12} sm={6}>
                        <h4>Modal Banner w/Title</h4>
                        <Button btnStyle="primary" onClick={this.toggle}>Show Modal</Button>
                        <Modal isOpen={this.state.modal} toggle={this.toggle}>
                            <ModalHeader toggle={this.toggle} />
                            <ModalBanner title="Add to Roster"/>
                            <ModalBody>
                                Good news, everyone! I've taught the toaster to feel love!
                            </ModalBody>
                            <ModalFooter>
                                <Button btnStyle="primary" onClick={this.toggle}>Do Something</Button>{' '}
                                <Button onClick={this.toggle}>Cancel</Button>
                            </ModalFooter>
                        </Modal>
                    </Col>
                    <Col xs={12} sm={6}>
                        <h4>Modal Banner w/Title + Indicator</h4>
                        <Button btnStyle="primary" onClick={this.toggleIndicator}>Show Modal</Button>
                        <Modal isOpen={this.state.modalIndicator} toggle={this.toggleIndicator}>
                            <ModalHeader toggle={this.toggleIndicator} />
                            <ModalBanner title="Add to Roster" indicator="add"/>
                            <ModalBody>
                                Good news, everyone! I've taught the toaster to feel love!
                            </ModalBody>
                            <ModalFooter>
                                <Button btnStyle="primary" onClick={this.toggleIndicator}>Do Something</Button>{' '}
                                <Button onClick={this.toggleIndicator}>Cancel</Button>
                            </ModalFooter>
                        </Modal>
                    </Col>
                </Row>
                <Row className={this.props.className}>
                    <Col xs={12}>
                        <h4>Modal Banner w/Title + Indicator + Subtitle</h4>
                        <Button btnStyle="primary" onClick={this.toggleSubtitle}>Show Modal</Button>
                        <Modal isOpen={this.state.modalSubtitle} toggle={this.toggleSubtitle}>
                            <ModalHeader toggle={this.toggleSubtitle} />
                            <ModalBanner title="Add to Roster" indicator="add">
                                <SecondaryFont>
                                    <Row>
                                        <Col>
                                            <span>Yoga for Beginners</span>
                                            <VerticalDivider />
                                            <span>8/21 8am</span>
                                        </Col>
                                    </Row>
                                </SecondaryFont>
                            </ModalBanner>
                            <ModalBody>
                                Good news, everyone! I've taught the toaster to feel love!
                            </ModalBody>
                            <ModalFooter>
                                <Button btnStyle="primary" onClick={this.toggleSubtitle}>Do Something</Button>{' '}
                                <Button onClick={this.toggleSubtitle}>Cancel</Button>
                            </ModalFooter>
                        </Modal>
                    </Col>
                </Row>
            </div>
        )
    }
}

ModalBannerExample.propTypes = {
    className: PropTypes.string,
}
