// @abc/fedagents/src/components/banners/modalBanner/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Indicator} from '@abc/protonpack'
import styles from './modalBanner.styl'
import classnames from 'classnames'

export default class ModalBanner extends Component {
    render() {
        const {indicator, title, children, className, styleType, ...remain} = this.props

        const classes = classnames(styles.container,className)

        return (
            <div className={classes} data-abc-id="modalBanner" {...remain}>
                {indicator &&
                    <Indicator type={indicator} className={styles.indicator} styleType={styleType} />
                }
                <h1 className={styles.title} >{title}</h1>
                <div data-abc-id="modalBannerContent" className={styles.bannerContent}>
                    {children}
                </div>
            </div>
        )
    }
}

ModalBanner.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    indicator: PropTypes.string,
    styleType: PropTypes.string,
    title: PropTypes.string.isRequired,
}
