// @abc/protonpack/src/components/image/userImage/example.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Row, Col, UserImage, SecondaryFont} from '../../../export'

export default class UserImageExample extends Component {
    render() {
        const colStyleLight = {
            alignItems: 'center',
            background:'#fff',
            color: '#4a4a4a',
            display: 'flex',
            flexDirection: 'column',
            justifyContent: 'flex-start',
            padding: '1em 0',
            textAlign: 'center',
        }
        const colStyleDark = {
            alignItems: 'center',
            background:'#4a4a4a',
            color: '#fff',
            display: 'flex',
            flexDirection: 'column',
            justifyContent: 'flex-start',
            padding: '1em 0',
            textAlign: 'center',
        }
        return (
            <div className="exampleSection">
                <h3 className="mb-3">User Image</h3>
                <Row>
                    <Col style={colStyleLight}>
                        <UserImage src="http://imgur.com/a1sLRv4.png" size="4.375rem"/>
                        <div>Default Color</div>
                    </Col>
                    <Col style={colStyleLight}>
                        <UserImage src="./bogus.png" size="4.375rem" hasBorder/>
                        <div>Default Color - w/Border</div>
                    </Col>
                    <Col style={colStyleLight}>
                        <div>
                            <SecondaryFont color="primary">
                                <UserImage src="./bogus.png" size="4.375rem" inheritColor/>
                            </SecondaryFont>
                            <div>Inherit Color</div>
                        </div>
                    </Col>
                    <Col style={colStyleLight}>
                        <div>
                            <SecondaryFont color="primary">
                                <UserImage src="./bogus.png" size="4.375rem" inheritColor hasBorder/>
                            </SecondaryFont>
                            <div>Inherit Color - w/Border</div>
                        </div>
                    </Col>
                </Row>
                <Row>
                    <Col style={colStyleDark}>
                        <UserImage src="./bogus.png" size="4.375rem"/>
                        <div>Default Color</div>
                    </Col>
                    <Col style={colStyleDark}>
                        <UserImage src="./bogus.png" size="4.375rem" hasBorder/>
                        <div>Default Color - w/Border</div>
                    </Col>
                    <Col style={colStyleDark}>
                        <div>
                            <SecondaryFont color="primary">
                                <UserImage src="./bogus.png" size="4.375rem" inheritColor/>
                            </SecondaryFont>
                            <div>Inherit Color</div>
                        </div>
                    </Col>
                    <Col style={colStyleDark}>
                        <div>
                            <SecondaryFont color="primary">
                                <UserImage src="./bogus.png" size="4.375rem" inheritColor hasBorder/>
                            </SecondaryFont>
                            <div>Inherit Color - w/Border</div>
                        </div>
                    </Col>
                </Row>
                <Row>
                    <Col style={colStyleLight}>
                        <UserImage src="http://imgur.com/a1sLRv4.png" alt="John A. Zoidberg" backgroundColor="darkGray"/>
                        <div>With Image</div>
                    </Col>
                    <Col style={colStyleLight}>
                        <UserImage src="" size="4.375rem" alt="John A. Zoidberg" backgroundColor="darkGray"/>
                        <div>With Initials</div>
                    </Col>
                    <Col style={colStyleLight}>
                        <UserImage src="http://imgur.com/a1sLRv4.png" size="4.375rem" alt="John A. Zoidberg" hasBorder  backgroundColor="darkGray"/>
                        <div>With Image and Border</div>
                    </Col>
                    <Col style={colStyleLight}>
                        <UserImage src="" size="4.375rem" alt="John A. Zoidberg" hasBorder backgroundColor="darkGray"/>
                        <div>With Initials and Border</div>
                    </Col>
                </Row>
            </div>
        )
    }
}

UserImageExample.propTypes = {
    className: PropTypes.string,
}
