// @abc/protonpack/src/components/image/userImage/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import BalancedImage, {colorList} from '../balancedImage'
import UserImageSVG from './userImageSVG'
import {Row, Col, SecondaryFont} from '../../../export'

export default class UserImage extends Component {
    renderInitials(alternateText) {
        const rowStyle = {
            height: '100%',
        }
        return (
            <Row style={rowStyle} xs={{justifyContent: 'center'}}>
                <Col xs={{alignSelf: 'center'}}>
                    <SecondaryFont color='white' size='xl'>
                        <span>{`${alternateText.charAt(0)}`}</span>
                    </SecondaryFont>
                </Col>
            </Row>
        )
    }

    render() {
        const {src, size, backgroundColor, alt, ...remain} = this.props
        const imageSrc = src
        const imageSize = size || '4.375rem'
        const placeholder = alt ? this.renderInitials(alt) : (<UserImageSVG />)

        return (
            <BalancedImage
                src={imageSrc}
                placeholder={placeholder}
                size={imageSize}
                backgroundColor={backgroundColor}
                round
                data-abc-id="userImage"
                alt={alt}
                {...remain}
            />
        )
    }
}

UserImage.propTypes = {
    src: PropTypes.string,
    size: PropTypes.string,
    alt: PropTypes.string,
    backgroundColor: PropTypes.oneOf(colorList),
}

UserImage.defaultProps = {
    src: '',
}
