// @abc/protonpack/src/components/formElements/toggleButton/example.js
import React, {Component} from 'react'
import {ToggleButton} from '../export'
import Icon from '@abc/quarkicons'
import {BaseText, Row, Col} from '../../../export'

export default class ToggleButtonExample extends Component {
    constructor(props) {
        super(props)

        this.state = {}

        this.handleBlur = this.handleBlur.bind(this)
        this.handleChange = this.handleChange.bind(this)
        this.handleFocus = this.handleFocus.bind(this)
    }

    handleBlur(toggleId) {
        this.setState({[`${toggleId}_focused`]: false})
    }

    handleChange(toggleId, value) {
        this.setState({[`${toggleId}_checked`]: value})
    }

    handleFocus(toggleId) {
        this.setState({[`${toggleId}_focused`]: true})
    }

    render() {
        return (
            <div className="exampleSection">
                <h2 className="mb-4">Toggle Buttons</h2>
                <Row className="mb-2">
                    <Col xs={12}>
                        <h4>Toggle Button - Interactive</h4>
                    </Col>
                    <Col sm={6}>
                        <div className="pt-2 pb-2">
                            <h5>Standard</h5>
                            <ToggleButton
                                checked={this.state.standard_checked || false}
                                onChange={value => this.handleChange('standard', value)}
                                onBlur={() => this.handleBlur('standard')}
                                onFocus={() => this.handleFocus('standard')}
                            >
                                <div>
                                    { this.state.standard_checked &&
                                        <b><Icon type="checkmark" /> Checked</b>
                                    }
                                    { !this.state.standard_checked &&
                                        <b><Icon type="x" /> Unchecked</b>
                                    }
                                    { this.state.standard_focused &&
                                        <span> and Focused</span>
                                    }
                                    { !this.state.standard_focused &&
                                        <span> and Unfocused</span>
                                    }
                                </div>
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6}>
                        <div className="pt-2 pb-2">
                            <h5>Standard Active</h5>
                            <ToggleButton
                                className="active"
                                checked={this.state.standardActive_checked || false}
                                onChange={value => this.handleChange('standardActive', value)}
                                onBlur={() => this.handleBlur('standardActive')}
                                onFocus={() => this.handleFocus('standardActive')}
                            >
                                <div>
                                    { this.state.standardActive_checked &&
                                        <b><Icon type="checkmark" /> Active Checked</b>
                                    }
                                    { !this.state.standardActive_checked &&
                                        <b><Icon type="x" /> Active Unchecked</b>
                                    }
                                    { this.state.standardActive_focused &&
                                        <span> and Focused</span>
                                    }
                                    { !this.state.standardActive_focused &&
                                        <span> and Unfocused</span>
                                    }
                                </div>
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6}>
                        <div className="pt-2 pb-2">
                            <h5>Flat</h5>
                            <ToggleButton
                                flat
                                checked={this.state.flat_checked || false}
                                onChange={value => this.handleChange('flat', value)}
                                onBlur={() => this.handleBlur('flat')}
                                onFocus={() => this.handleFocus('flat')}
                            >
                                <div>
                                    { this.state.flat_checked &&
                                        <b><Icon type="checkmark" /> Checked</b>
                                    }
                                    { !this.state.flat_checked &&
                                        <b><Icon type="x" /> Unchecked</b>
                                    }
                                    { this.state.flat_focused &&
                                        <span> and Focused</span>
                                    }
                                    { !this.state.flat_focused &&
                                        <span> and Unfocused</span>
                                    }
                                </div>
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6}>
                        <div className="pt-2 pb-2">
                            <h5>Flat Active</h5>
                            <ToggleButton
                                flat
                                className="active"
                                checked={this.state.flatActive_checked || false}
                                onChange={value => this.handleChange('flatActive', value)}
                                onBlur={() => this.handleBlur('flatActive')}
                                onFocus={() => this.handleFocus('flatActive')}
                            >
                                <div>
                                    { this.state.flatActive_checked &&
                                        <b><Icon type="checkmark" /> Active Checked</b>
                                    }
                                    { !this.state.flatActive_checked &&
                                        <b><Icon type="x" /> Active Unchecked</b>
                                    }
                                    { this.state.flatActive_focused &&
                                        <span> and Focused</span>
                                    }
                                    { !this.state.flatActive_focused &&
                                        <span> and Unfocused</span>
                                    }
                                </div>
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6}>
                        <div className="pt-2 pb-2">
                            <h5>Flat Outline</h5>
                            <ToggleButton
                                flat
                                outline
                                checked={this.state.flatOutline_checked || false}
                                onChange={value => this.handleChange('flatOutline', value)}
                                onBlur={() => this.handleBlur('flatOutline')}
                                onFocus={() => this.handleFocus('flatOutline')}
                            >
                                <div>
                                    { this.state.flatOutline_checked &&
                                        <b><Icon type="checkmark" /> Checked</b>
                                    }
                                    { !this.state.flatOutline_checked &&
                                        <b><Icon type="x" /> Unchecked</b>
                                    }
                                    { this.state.flatOutline_focused &&
                                        <span> and Focused</span>
                                    }
                                    { !this.state.flatOutline_focused &&
                                        <span> and Unfocused</span>
                                    }
                                </div>
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6}>
                        <div className="pt-2 pb-2">
                            <h5>Flat Outline Active</h5>
                            <ToggleButton
                                flat
                                outline
                                className="active"
                                checked={this.state.flatOutlineActive_checked || false}
                                onChange={value => this.handleChange('flatOutlineActive', value)}
                                onBlur={() => this.handleBlur('flatOutlineActive')}
                                onFocus={() => this.handleFocus('flatOutlineActive')}
                            >
                                <div>
                                    { this.state.flatOutlineActive_checked &&
                                        <b><Icon type="checkmark" /> Active Checked</b>
                                    }
                                    { !this.state.flatOutlineActive_checked &&
                                        <b><Icon type="x" /> Active Unchecked</b>
                                    }
                                    { this.state.flatOutlineActive_focused &&
                                        <span> and Focused</span>
                                    }
                                    { !this.state.flatOutlineActive_focused &&
                                        <span> and Unfocused</span>
                                    }
                                </div>
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6} md={3}>
                        <div className="pt-2 pb-2">
                            <h5>Round</h5>
                            <ToggleButton
                                round
                                checked={this.state.round_checked || false}
                                onChange={value => this.handleChange('round', value)}
                                onBlur={() => this.handleBlur('round')}
                                onFocus={() => this.handleFocus('round')}
                            >
                                { this.state.round_checked &&
                                    <Icon type="minus" />
                                }
                                { !this.state.round_checked &&
                                    <Icon type="plus" />
                                }
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6} md={3}>
                        <div className="pt-2 pb-2">
                            <h5>Round Active</h5>
                            <ToggleButton
                                round
                                className="active"
                                checked={this.state.roundActive_checked || false}
                                onChange={value => this.handleChange('roundActive', value)}
                                onBlur={() => this.handleBlur('roundActive')}
                                onFocus={() => this.handleFocus('roundActive')}
                            >
                                { this.state.roundActive_checked &&
                                    <Icon type="minus" />
                                }
                                { !this.state.roundActive_checked &&
                                    <Icon type="plus" />
                                }
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6} md={3}>
                        <div className="pt-2 pb-2">
                            <h5>Round Flat</h5>
                            <ToggleButton
                                round
                                flat
                                checked={this.state.roundFlat_checked || false}
                                onChange={value => this.handleChange('roundFlat', value)}
                                onBlur={() => this.handleBlur('roundFlat')}
                                onFocus={() => this.handleFocus('roundFlat')}
                            >
                                { this.state.roundFlat_checked &&
                                    <BaseText size="md"><span>Full</span></BaseText>
                                }
                                { !this.state.roundFlat_checked &&
                                    <BaseText size="md"><span>None</span></BaseText>
                                }
                            </ToggleButton>
                        </div>
                    </Col>
                    <Col sm={6} md={3}>
                        <div className="pt-2 pb-2">
                            <h5>Round Flat Active</h5>
                            <ToggleButton
                                round
                                flat
                                className="active"
                                checked={this.state.roundFlatActive_checked || false}
                                onChange={value => this.handleChange('roundFlatActive', value)}
                                onBlur={() => this.handleBlur('roundFlatActive')}
                                onFocus={() => this.handleFocus('roundFlatActive')}
                            >
                                { this.state.roundFlatActive_checked &&
                                    <BaseText size="md"><span>Full</span></BaseText>
                                }
                                { !this.state.roundFlatActive_checked &&
                                    <BaseText size="md"><span>None</span></BaseText>
                                }
                            </ToggleButton>
                        </div>
                    </Col>
                </Row>
                <Row className="mb-2">
                    <Col xs={12} sm={6} lg={3}>
                        <h4>Toggle Button - Default</h4>
                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton className="float-left"/>
                                Unchecked
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton checked className="float-left"/>
                                Checked
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton disabled className="float-left"/>
                                Disabled
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton checked disabled className="float-left"/>
                                Disabled + Checked
                            </label>
                        </div>
                    </Col>
                    <Col xs={12} sm={6} lg={3}>
                        <h4>Toggle Button - w/Content</h4>
                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton className="float-left">
                                    <span>Toggle Content</span>
                                </ToggleButton>
                                Unchecked
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton checked className="float-left">
                                    <span>Toggle Content</span>
                                </ToggleButton>
                                Checked
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton disabled className="float-left">
                                    <span>Toggle Content</span>
                                </ToggleButton>
                                Disabled
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton checked disabled className="float-left">
                                    <span>Toggle Content</span>
                                </ToggleButton>
                                Disabled + Checked
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton flat className="float-left">
                                    <span>Toggle Content</span>
                                </ToggleButton>
                                Flat + Unchecked
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton flat checked className="float-left">
                                    <span>Toggle Content</span>
                                </ToggleButton>
                                Flat + Checked
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton flat disabled className="float-left">
                                    <span>Toggle Content</span>
                                </ToggleButton>
                                Flat + Disabled
                            </label>
                        </div>

                        <div className="pt-2 pb-2">
                            <label>
                                <ToggleButton flat checked disabled className="float-left">
                                    <span>Toggle Content</span>
                                </ToggleButton>
                                Flat + Disabled + Checked
                            </label>
                        </div>
                    </Col>
                    <Col xs={12} sm={6} lg={3}>
                        <h4>Toggle Button - Sizes</h4>
                        <div className="pt-2 pb-2">
                            <ToggleButton size={'xs'}/>
                            <ToggleButton size={'xs'} checked/>
                            <ToggleButton size={'xs'} disabled/>
                            <ToggleButton size={'xs'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton size={'sm'}/>
                            <ToggleButton size={'sm'} checked/>
                            <ToggleButton size={'sm'} disabled/>
                            <ToggleButton size={'sm'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton size={'md'}/>
                            <ToggleButton size={'md'} checked/>
                            <ToggleButton size={'md'} disabled/>
                            <ToggleButton size={'md'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton size={'lg'}/>
                            <ToggleButton size={'lg'} checked/>
                            <ToggleButton size={'lg'} disabled/>
                            <ToggleButton size={'lg'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton size={'xl'}/>
                            <ToggleButton size={'xl'} checked/>
                            <ToggleButton size={'xl'} disabled/>
                            <ToggleButton size={'xl'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton size={'xxl'}/>
                            <ToggleButton size={'xxl'} checked/>
                            <ToggleButton size={'xxl'} disabled/>
                            <ToggleButton size={'xxl'} checked disabled/>
                        </div>
                    </Col>
                    <Col xs={12} sm={6} lg={3}>
                        <h4>Toggle Button - Colors</h4>
                        <div className="pt-2 pb-2">
                            <ToggleButton btnStyle={'primary'}/>
                            <ToggleButton btnStyle={'primary'} checked/>
                            <ToggleButton btnStyle={'primary'} disabled/>
                            <ToggleButton btnStyle={'primary'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton btnStyle={'secondary'}/>
                            <ToggleButton btnStyle={'secondary'} checked/>
                            <ToggleButton btnStyle={'secondary'} disabled/>
                            <ToggleButton btnStyle={'secondary'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton btnStyle={'info'}/>
                            <ToggleButton btnStyle={'info'} checked/>
                            <ToggleButton btnStyle={'info'} disabled/>
                            <ToggleButton btnStyle={'info'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton btnStyle={'disabled'}/>
                            <ToggleButton btnStyle={'disabled'} checked/>
                            <ToggleButton btnStyle={'disabled'} disabled/>
                            <ToggleButton btnStyle={'disabled'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton btnStyle={'warning'}/>
                            <ToggleButton btnStyle={'warning'} checked/>
                            <ToggleButton btnStyle={'warning'} disabled/>
                            <ToggleButton btnStyle={'warning'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton btnStyle={'success'}/>
                            <ToggleButton btnStyle={'success'} checked/>
                            <ToggleButton btnStyle={'success'} disabled/>
                            <ToggleButton btnStyle={'success'} checked disabled/>
                        </div>

                        <div className="pt-2 pb-2">
                            <ToggleButton btnStyle={'danger'}/>
                            <ToggleButton btnStyle={'danger'} checked/>
                            <ToggleButton btnStyle={'danger'} disabled/>
                            <ToggleButton btnStyle={'danger'} checked disabled/>
                        </div>
                    </Col>
                    <Col xs={12} sm={6}>
                        <h4>Round Toggle Buttons</h4>
                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton round className="float-left">
                                <BaseText size="md"><span>Full</span></BaseText>
                            </ToggleButton>
                            <span>Round + Unchecked</span>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton round checked className="float-left">
                                <BaseText size="md"><span>Full</span></BaseText>
                            </ToggleButton>
                            <span>Round + Checked</span>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton round disabled className="float-left">
                                <BaseText size="md"><span>Full</span></BaseText>
                            </ToggleButton>
                            <span>Round + Disabled</span>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton round checked disabled className="float-left">
                                <BaseText size="md"><span>Full</span></BaseText>
                            </ToggleButton>
                            <span>Round + Disabled + Checked</span>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round className="float-left">
                                <BaseText size="md"><span>Full</span></BaseText>
                            </ToggleButton>
                            <span>Flat + Round + Unchecked</span>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round checked className="float-left">
                                <BaseText size="md"><span>Full</span></BaseText>
                            </ToggleButton>
                            <span>Flat + Round + Checked</span>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round disabled className="float-left">
                                <BaseText size="md"><span>Full</span></BaseText>
                            </ToggleButton>
                            <span>Flat + Round + Disabled</span>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round checked disabled className="float-left">
                                <BaseText size="md"><span>Full</span></BaseText>
                            </ToggleButton>
                            <span>Flat + Round + Disabled + Checked</span>
                        </div>
                    </Col>
                    <Col xs={12} sm={6}>
                        <h4>Flat Round Toggle Buttons - Colors</h4>
                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round btnStyle={'primary'}><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'primary'} checked><Icon type="plus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'primary'} disabled><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'primary'} checked disabled><Icon type="plus" /></ToggleButton>
                            <code>btnStyle="primary"</code>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round btnStyle={'secondary'}><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'secondary'} checked><Icon type="plus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'secondary'} disabled><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'secondary'} checked disabled><Icon type="plus" /></ToggleButton>
                            <code>btnStyle="secondary"</code>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round btnStyle={'info'}><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'info'} checked><Icon type="plus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'info'} disabled><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'info'} checked disabled><Icon type="plus" /></ToggleButton>
                            <code>btnStyle="info"</code>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round btnStyle={'disabled'}><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'disabled'} checked><Icon type="plus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'disabled'} disabled><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'disabled'} checked disabled><Icon type="plus" /></ToggleButton>
                            <code>btnStyle="disabled"</code>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round btnStyle={'warning'}><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'warning'} checked><Icon type="plus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'warning'} disabled><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'warning'} checked disabled><Icon type="plus" /></ToggleButton>
                            <code>btnStyle="warning"</code>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round btnStyle={'success'}><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'success'} checked><Icon type="plus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'success'} disabled><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'success'} checked disabled><Icon type="plus" /></ToggleButton>
                            <code>btnStyle="success"</code>
                        </div>

                        <div className="pt-2 pb-2" style={{display: 'flex', alignItems: 'center'}}>
                            <ToggleButton flat round btnStyle={'danger'}><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'danger'} checked><Icon type="plus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'danger'} disabled><Icon type="minus" /></ToggleButton>
                            <ToggleButton flat round btnStyle={'danger'} checked disabled><Icon type="plus" /></ToggleButton>
                            <code>btnStyle="danger"</code>
                        </div>
                    </Col>
                </Row>
            </div>
        )
    }
}
