// @abc/protonpack/src/components/formElements/reactSelect/example.js
import React, {Component} from 'react'
import {
    FormGroup,
    Label,
    ReactSelect,
    Row,
    Col,
} from 'src/export'

export default class ReactSelectExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            reactMultiSelectValue: '',
            creatableSelectValue: '',
            reactSelectValue: '',
            editableSelectValue: '',
        }
        this.farnsworths = [
            {value: 'professor', label: 'Professor Farnsworth'},
            {value: 'cubert', label: 'Cubert Farnsworth'},
            {value: 'philo', label: 'Philo Farnsworth'},
            {value: 'david', label: 'David Farnsworth'},
            {value: 'floyd', label: 'Floyd Farnsworth'},
            {value: 'ned', label: 'Ned Farnsworth'},
            {value: 'velma', label: 'Velma Farnsworth'},
            {value: 'fry', label: 'Phillip J. Fry'},
            {value: 'igner', label: 'Igner'},
        ]

        this.creatableFarnsworths = [...this.farnsworths]
        this.editableFarnsworths = [...this.farnsworths]

        this.handleChange = this.handleChange.bind(this)
        this.handleMultiSelectChange = this.handleMultiSelectChange.bind(this)
        this.handleCreatableSelectChange = this.handleCreatableSelectChange.bind(this)
        this.handleEditableSelectChange = this.handleEditableSelectChange.bind(this)
    }

    handleChange(newValue) {
        this.setState({reactSelectValue: newValue})
    }

    handleMultiSelectChange(newValue) {
        this.setState({reactMultiSelectValue: newValue})
    }

    handleCreatableSelectChange(newValue) {
        this.setState({creatableSelectValue: newValue})
    }

    handleEditableSelectChange(newValue) {
        this.setState({editableSelectValue: newValue})
    }

    render() {
        return (
            <div className="exampleSection">
                <h4 className="mb-2">ReactSelect</h4>
                <Row className="mb-2">
                    <Col xs={4}>
                        <FormGroup>
                            <Label>Basic</Label>
                            <ReactSelect
                                options={this.farnsworths}
                                value={this.state.reactSelectValue}
                                onChange={this.handleChange}
                            />
                        </FormGroup>
                    </Col>
                    <Col xs={4}>
                        <FormGroup>
                            <Label>Disabled</Label>
                            <ReactSelect
                                options={this.farnsworths}
                                value={this.state.reactSelectValue}
                                onChange={this.handleChange}
                                disabled
                            />
                        </FormGroup>
                    </Col>
                    <Col xs={4}>
                        <FormGroup>
                            <Label>Multi</Label>
                            <ReactSelect
                                simpleValue
                                placeholder="Pick a Farnsworth"
                                value={this.state.reactMultiSelectValue}
                                onChange={this.handleMultiSelectChange}
                                onBlur={this.handleMultiSelectChange}
                                options={this.farnsworths}
                                multi
                                backspaceRemoves
                            />
                        </FormGroup>
                    </Col>
                    <Col xs={4}>
                        <FormGroup>
                            <Label>Creatable</Label>
                            <ReactSelect
                                simpleValue
                                placeholder="Pick a Farnsworth"
                                value={this.state.creatableSelectValue}
                                onChange={this.handleCreatableSelectChange}
                                onBlur={this.handleCreatableSelectChange}
                                options={this.creatableFarnsworths}
                                multi
                                creatable
                                backspaceRemoves
                            />
                        </FormGroup>
                    </Col>
                    <Col xs={4}>
                        <FormGroup>
                            <Label>Editable</Label>
                            <ReactSelect
                                simpleValue
                                placeholder="Pick a Farnsworth"
                                value={this.state.editableSelectValue}
                                onChange={this.handleEditableSelectChange}
                                onBlur={this.handleEditableSelectChange}
                                options={this.editableFarnsworths}
                                editable
                            />
                        </FormGroup>
                    </Col>
                </Row>
            </div>
        )
    }
}
