// @abc/protonpack/src/brander/button/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {Row, Col} from 'layout/export'
import ColorPicker from '../colorPicker'
import {FormGroup, LabelGroup, ReactSelect, Label} from 'formElements/export'
import BranderRange from '../common/range/BranderRange'
import {styleTypes, elementStates} from '../data'
import './buttonBrander.scss'

class ButtonBrander extends Component {
    constructor(props) {
        super(props)

        this.state = {
            styleType: '',
            elementState: '',
        }
    }

    render() {
        const {styles} = this.props
        const {styleType, elementState} = this.state

        return (
            <Row xs={{justifyContent: 'center'}}>
                <Col sm={12}>
                    <FormGroup size="sm" >
                        <LabelGroup label="Border Radius" standaloneLabel >
                            <BranderRange
                                id="borderRadius"
                                CSSProperty="--buttonBorderRadius"
                                styles={styles}
                                min={0}
                                max={30}
                            />
                        </LabelGroup>
                    </FormGroup>
                </Col>
                <Col sm={6}>
                    <FormGroup size="sm">
                        <Label>Style Type</Label>
                        <ReactSelect
                            size="sm"
                            options={styleTypes}
                            onChange={styleType => this.setState({styleType})}
                            value={this.state.styleType}
                            placeholder="Select One..."
                        />
                    </FormGroup>
                </Col>
                <Col sm={6}>
                    <FormGroup size="sm">
                        <Label>State</Label>
                        <ReactSelect
                            size="sm"
                            options={elementStates}
                            onChange={elementState => this.setState({elementState})}
                            value={this.state.elementState}
                        />
                    </FormGroup>
                </Col>
                {styleType &&
                    <Col sm={12}>
                        <Row>
                            <Col sm={12}>
                                <ColorPicker
                                    customProperty={`--button${styleType}${elementState}Background`}
                                    label="Background"
                                    className={`button-brander-${styleType.toLowerCase()}${elementState ? `-${elementState.toLowerCase()}-` : '-'}background`}
                                    color={String(styles.getPropertyValue(`--button${styleType}${elementState}Background`)).trim()}
                                    isCSSProperty
                                />
                            </Col>
                            <Col sm={6}>
                                <ColorPicker
                                    customProperty={`--button${styleType}${elementState}Text`}
                                    label="Text"
                                    className={`button-brander-${styleType.toLowerCase()}${elementState ? `-${elementState.toLowerCase()}-` : '-'}text`}
                                    color={String(styles.getPropertyValue(`--button${styleType}${elementState}Text`)).trim()}
                                />
                            </Col>
                            <Col sm={6}>
                                <ColorPicker
                                    customProperty={`--button${styleType}${elementState}Border`}
                                    label="Border Color"
                                    className={`button-brander-${styleType.toLowerCase()}${elementState ? `-${elementState.toLowerCase()}-` : '-'}border`}
                                    color={String(styles.getPropertyValue(`--button${styleType}${elementState}Border`)).trim()}
                                />
                            </Col>
                        </Row>
                    </Col>
                }
            </Row>
        )
    }
}

ButtonBrander.propTypes = {
    styles: PropTypes.object.isRequired,
}

export default ButtonBrander
