// @abc/protonpack/src/components/buttons/buttonGroup/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'

export default class ButtonGroup extends Component {
    buildClasses(props) {
        const classes = []

        if (props.vertical) {
            classes.push('btn-group-vertical')
        } else {
            classes.push('btn-group')
        }

        if (props.size) {
            classes.push(`btn-group-${props.size}`)
        }

        if (props.className) {
            classes.push(props.className)
        }

        return classes.join(' ')
    }

    render() {
        const {vertical, size, className, ...remain} = this.props
        const props = {vertical, size, className}

        return (
            <div className={this.buildClasses(props)} data-abc-id="buttonGroup" {...remain} >
                {this.props.children}
            </div>
        )
    }
}

ButtonGroup.propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    vertical: PropTypes.bool,
    size: PropTypes.string,
}
