// @abc/protonpack/src/components/signature/example.js
import React, {Component} from 'react'
import {SignaturePad, Row, Col, FormGroup, LabelGroup, Button, Checkbox} from '../../export'

export default class SignatureExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            signature: null,
            validateSignature: false,
            statusMessage: {},
        }

        this.handleSignaturePadSave = this.handleSignaturePadSave.bind(this)
    }

    handleSignaturePadSave() {
        if (!this.signaturePad.validate()) {
            console.warn('Must provide a signature first')
            this.setState({signature: null})
        } else {
            this.setState({signature: this.signaturePad.toDataURL()})
        }
    }

    render() {
        const {validateSignature} = this.state

        return (
            <div className="exampleSection">
                <h2 className="mb-2">Signature</h2>
                <Row>
                    <Col xs={{col:12, justifyContent: 'center'}} md={5}>
                        <LabelGroup label='Validate' standaloneLabel check inline>
                            <Checkbox
                                checked={validateSignature}
                                onChange={() => this.setState({validateSignature: !validateSignature})}
                            />
                        </LabelGroup>
                        <FormGroup>
                            <LabelGroup label='Sign' hideLabel>
                                <SignaturePad
                                    continuousValidation={validateSignature}
                                    ref={ref => this.signaturePad = ref}
                                    width={250}
                                    height={108}
                                />
                            </LabelGroup>
                        </FormGroup>
                        <Button onClick={this.handleSignaturePadSave}>Save</Button>
                    </Col>

                    {this.state.signature &&
                        <Col>
                            <h4>Preview Image</h4>
                            <img src={this.state.signature} style={{width: 250, backgroundColor: 'white'}}/>
                        </Col>
                    }
                </Row>
            </div>
        )
    }
}
