// @abc/protonpack/src/components/formElements/datepicker/example.js
import React, {Component} from 'react'
import {Col, DatePicker, Indicator, Row} from '../../../export'
import moment from 'moment'
import 'react-dates/lib/css/_datepicker.css'

class DatePickerExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            dateCenter: moment(),
            dateLeft: moment(),
            dateRight: moment(),
        }
    }

    isDayHighlighted(day) {
        const isInclusionDay = day.isSame(moment().add(2, 'days'), 'day')
        const isInInclusionDayRange = day.isBetween(moment().add(7, 'days'), 'day', moment().add(12, 'days'), 'day')

        return isInclusionDay || isInInclusionDayRange
    }

    render() {
        return (
            <Row>
                <Col xs={12} style={{textAlign: 'center'}}>
                    <DatePicker
                        date={this.state.dateCenter}
                        onDateChange={date => this.setState({dateCenter: date})}
                        display
                        allowPastDates
                        isDayHighlighted={this.isDayHighlighted}
                        data-abc-id="myTestId-center"
                        id="datePickerExample1"
                    />
                </Col>
                <Col xs={6} style={{textAlign: 'left'}}>
                    <DatePicker
                        date={this.state.dateLeft}
                        onDateChange={date => this.setState({dateLeft: date})}
                        display
                        allowPastDates
                        isDayHighlighted={this.isDayHighlighted}
                        data-abc-id="myTestId-center"
                        hideDisplayText
                        popoverAlign="left"
                        icon={<Indicator type="select-date"/>}
                        id="datePickerExample2"
                    />
                </Col>
                <Col xs={6} style={{textAlign: 'right'}}>
                    <DatePicker
                        date={this.state.dateRight}
                        onDateChange={date => this.setState({dateRight: date})}
                        display
                        allowPastDates
                        isDayHighlighted={this.isDayHighlighted}
                        data-abc-id="myTestId-center"
                        hideDisplayText
                        popoverAlign="right"
                        icon="chevron-down-o"
                        id="datePickerExample3"
                    />
                </Col>
            </Row>
        )
    }
}

export default DatePickerExample
