// @abc/protonpack/webpack.prod.config.js
const ExtractTextPlugin = require('extract-text-webpack-plugin')
const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin')
const path = require('path')

module.exports = {
    bail: true,
    entry: {
        protonpack: './src/export.js',
    },
    output: {
        path: path.join(__dirname, 'dist'),
        filename: '[name].js',
        library: 'protonpack',
        libraryTarget: 'umd',
    },
    externals: [
        {
            react: {
                root: 'React',
                commonjs2: 'react',
                commonjs: 'react',
                amd: 'react',
            },
            'react-dom': {
                root: 'ReactDOM',
                commonjs2: 'react-dom',
                commonjs: 'react-dom',
                amd: 'react-dom',
            },
            '@abc/quarkicons': {
                root: '@abc/quarkicons',
                commonjs2: '@abc/quarkicons',
                commonjs: '@abc/quarkicons',
                amd: '@abc/quarkicons',
            },
        },
    ],
    rules: [
        {
            test: /\.css$/,
            use: ExtractTextPlugin.extract({
                fallback: 'style-loader',
                use: 'css-loader',
            }),
        },
        {
            test: /\.scss$/,
            use: ExtractTextPlugin.extract(['css-loader', 'sass-loader']),
        },
        {
            test: /\.js$/,
            exclude: /node_modules/,
            loader: 'babel-loader',
            query: {presets: ['es2015', 'react']},
        },
    ],
    plugins: [
        new ExtractTextPlugin('style.css'),
        new OptimizeCssAssetsPlugin(),
    ],
    devServer: {},
    devtool: '',
}
