// @abc/protonpack/src/brander/data.js
export const brandingOptions = [
    {value: 'body', label: 'Body'},
    {value: 'link', label: 'Link'},
    {value: 'button', label: 'Button'},
    {value: 'form', label: 'Form'},
    {value: 'input', label: 'Input'},
    {value: 'select', label: 'Select'},
    {value: 'checkbox', label: 'Checkbox'},
    {value: 'ufo', label: 'UFO'},
    {value: 'alert', label: 'Alert'},
    {value: 'modal', label: 'Modal'},
    {value: 'popover', label: 'Popover'},
    {value: 'tooltip', label: 'Tooltip'},
    {value: 'divider', label: 'Divider'},
    {value: 'card', label: 'Card'},
    {value: 'inlineHover', label: 'Inline Hover'},
    {value: 'datePicker', label: 'Date Picker'},
]

export const styleTypes = [
    {value: 'Primary', label: 'Primary'},
    {value: 'Secondary', label: 'Secondary'},
]

export const statuses = [
    {value: 'Success', label: 'Success'},
    {value: 'Warning', label: 'Warning'},
    {value: 'Danger', label: 'Danger'},
]

export const elementStates = [
    {value: '', label: 'Normal'},
    {value: 'Hover', label: 'Hover'},
    {value: 'Focus', label: 'Focus'},
    {value: 'Disabled', label: 'Disabled'},
]

export const formElements = [
    {value: 'Label', label: 'Label'},
]

export const datePickerDayTypes = [
    {value: '', label: 'Normal'},
    {value: 'Hover', label: 'Hover'},
    {value: 'Selected', label: 'Selected'},
    {value: 'Blocked', label: 'Blocked'},
    {value: 'Highlighted', label: 'Highlighted'},
]

export const calendarElements = [
    {value: 'calendar', label: 'Calendar'},
    {value: 'day', label: 'Day'},
]

export const labelStates = elementStates.filter(state => !['Hover', 'Focus'].includes(state.label))
export const inputStates = elementStates.filter(state => !['Hover'].includes(state.label))
