// @abc/fedagents/src/components/confirmModal/example.js
import React, {Component} from 'react'
import {Row, Col, Button} from '@abc/protonpack'
import {ConfirmModal} from '../../export'

export default class ConfirmModalExample extends Component {
    constructor(props) {
        super(props)

        this.state = {
            modalBase: false,
            modalError: false,
            errorMessage: '',
        }

        this.toggleBase = this.toggleBase.bind(this)
        this.handleBaseClickOk = this.handleBaseClickOk.bind(this)
        this.toggleError = this.toggleError.bind(this)
        this.handleErrorClickOk = this.handleErrorClickOk.bind(this)
        this.testCall = this.testCall.bind(this)
    }

    toggleBase() {
        this.setState({
            modalBase: !this.state.modalBase,
        })
    }

    handleBaseClickOk() {
        this.setState({isBaseModalSubmitting: true})

        return new Promise((resolve, reject) => {
            setTimeout(
                () => this.testCall(resolve, reject),
                3000
            )
        })
    }

    testCall(resolve, reject, fail = false) {
        let errorMessage = ''

        if (fail) {
            reject('failed')
            errorMessage = 'The call has failed. Please try again.'
        } else {
            resolve()
        }

        this.setState({
            modalBase: false,
            isBaseModalSubmitting: false,
            errorMessage,
        })
    }

    toggleError() {
        this.setState({
            modalError: !this.state.modalError,
            errorMessage: '',
        })
    }

    handleErrorClickOk() {
        this.setState({modalError: true})

        return new Promise((resolve, reject) => {
            setTimeout(
                () => this.testCall(resolve, reject, true),
                3000
            )
        })
    }

    render() {
        const {modalBase, modalError, errorMessage} = this.state

        return (
            <div className="exampleSection">
                <h2 className="mb-2">Confirmation Modal</h2>
                <Row>
                    <Col>
                        <Button btnStyle="primary" block  onClick={this.toggleBase}>Test Danger Modal</Button>
                        <ConfirmModal 
                            title="Danger, Will Robinson!" 
                            toggle={this.toggleBase}
                            onOk={this.handleBaseClickOk}
                            isOpen={modalBase}
                            okText="Cool"
                        >
                            You accidentally created a hole in space and time.
                        </ConfirmModal>
                    </Col>
                    <Col>
                        <Button btnStyle="primary" block  onClick={this.toggleError}>Test Danger Modal w/ Error Message</Button>
                        <ConfirmModal 
                            title="Danger, Will Robinson!" 
                            toggle={this.toggleError}
                            onOk={this.handleErrorClickOk}
                            isOpen={modalError}
                            okText="Cool"
                            errorMessage={errorMessage}
                        >
                            You accidentally created a hole in space and time.
                        </ConfirmModal>
                    </Col>
                </Row>
            </div>
        )
    }
}
