// @abc/protonpack/src/components/formElements/checkbox/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'

export default class Checkbox extends Component {
    buildClasses() {
        const {size, className, statusMessage} = this.props
        const classes = ['checkboxContainer']

        if (className) {
            classes.push(className)
        }
        if (size) {
            classes.push(`checkbox-size-${size}`)
        }

        if (statusMessage && statusMessage.status) {
            classes.push(`form-control-${statusMessage.status}`)
        }

        return classes.join(' ')
    }

    render() {
        // eslint-disable-next-line no-unused-vars
        const {['data-abc-id']: dataAbcId, size, className, statusMessage, ...formGroup} = this.props
        const classes = this.buildClasses()
        const abcId = dataAbcId || 'checkboxInput'

        // eslint-disable-next-line no-unused-vars
        const {normalize, validate, onValidationFieldInitialized, onValidationFieldFailure, onValidationFieldSuccess, ...remain} = formGroup

        return (
            <div className={classes}>
                <input type="checkbox" className={'form-check-input'} data-abc-id={abcId} {...remain}/>
                <span className="form-checkbox"/>
            </div>
        )
    }
}

Checkbox.propTypes = {
    className: PropTypes.string,
    size: PropTypes.string,
    statusMessage: PropTypes.object,
}
