// @abc/protonpack/src/components/_newComponentTemplate/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

export default class NewComponentTemplate extends Component {
    buildClasses(props) {
        // Use this method to preform any more advanced evaluations needed to provide
        // the classname library with values

        // Use the classname as the object key followed by a bool of 
        // whether or not it should be displayed
        // -- docs: https://github.com/JedWatson/classnames 
        return classnames(
            'row__example--flex', // default class always returned
            {junk: (props.someProp === 'junk')}, 
            {[`other--${props.someProp}`]: (props.someProp === 'otherStuff')},
            {'other--none': !props.someOtherProp}
        )
    }

    render() {
        // eslint-disable-next-line no-unused-vars
        const {someProp, someOtherProp, ['data-abc-id']: dataAbcId, ...remain} = this.props
        const abcId = dataAbcId || 'defaultAbcId'

        // Remember to spread remain on the top level tag to handle all uncaught HTML properties
        return (
            <div data-abc-id={abcId} className={this.buildClasses(this.props)} {...remain}>
                Some Example {someOtherProp}
            </div>
        )
    }
}

NewComponentTemplate.propTypes = {
    someProp: PropTypes.string,
    someOtherProp: PropTypes.string.isRequired,
}
